File: packages/python/src/telebot_py/types/gifts.py
"""Telegram gift, collectible-gift, and owned-gift types.

Field sets mirror the authoritative TypeScript sources
``packages/node/src/client/types/payments/models.ts`` (``Gift``, ``Gifts``,
``UniqueGift*``) extended with the owned-gift containers and
``AcceptedGiftTypes`` that the gifts Bot API methods return and accept.
"""

from __future__ import annotations

import dataclasses

from telebot_py.types.base import TelegramObject
from telebot_py.types.chat import Chat
from telebot_py.types.common import MessageEntity
from telebot_py.types.stickers import Sticker
from telebot_py.types.user import User


@dataclasses.dataclass(frozen=True, slots=True)
class Gift(TelegramObject):
    """This object represents a gift that can be sent by the bot.

    Attributes:
        id: Unique identifier of the gift.
        sticker: The sticker that represents the gift.
        star_count: Number of Telegram Stars that must be paid to send the
            sticker.
        upgrade_star_count: Number of Telegram Stars that must be paid to
            upgrade the gift to a unique one.
        total_count: The total number of the gifts of this type that can be
            sent; for limited gifts only.
        remaining_count: The number of remaining gifts of this type that can
            be sent; for limited gifts only.
    """

    id: str
    sticker: Sticker
    star_count: int
    upgrade_star_count: int | None = None
    total_count: int | None = None
    remaining_count: int | None = None


@dataclasses.dataclass(frozen=True, slots=True)
class Gifts(TelegramObject):
    """A list of gifts.

    Returned by ``Bot.get_available_gifts``.

    Attributes:
        gifts: The list of gifts.
    """

    gifts: list[Gift]


@dataclasses.dataclass(frozen=True, slots=True)
class AcceptedGiftTypes(TelegramObject):
    """Information about types of gifts accepted by a user or business account.

    Attributes:
        unlimited_gifts: True, if unlimited regular gifts are accepted.
        limited_gifts: True, if limited regular gifts are accepted.
        unique_gifts: True, if unique gifts or gifts that can be upgraded to
            unique for free are accepted.
        premium_subscription: True, if a Telegram Premium subscription is
            accepted.
        gifts_from_channels: True, if gifts from channels are accepted.
    """

    unlimited_gifts: bool
    limited_gifts: bool
    unique_gifts: bool
    premium_subscription: bool
    gifts_from_channels: bool


@dataclasses.dataclass(frozen=True, slots=True)
class UniqueGiftBackdropColors(TelegramObject):
    """Colors of a unique gift backdrop.

    Attributes:
        center_color: The color in the center of the backdrop in RGB format.
        edge_color: The color on the edges of the backdrop in RGB format.
        symbol_color: The color to be applied to the symbol in RGB format.
        text_color: The color for the text on the backdrop in RGB format.
    """

    center_color: int
    edge_color: int
    symbol_color: int
    text_color: int


@dataclasses.dataclass(frozen=True, slots=True)
class UniqueGiftBackdrop(TelegramObject):
    """The backdrop of a unique gift.

    Attributes:
        name: The name of the backdrop.
        colors: The colors of the backdrop.
        rarity_per_mille: The number of unique gifts that receive this backdrop
            for every 1000 gifts upgraded.
    """

    name: str
    colors: UniqueGiftBackdropColors
    rarity_per_mille: int


@dataclasses.dataclass(frozen=True, slots=True)
class UniqueGiftModel(TelegramObject):
    """The model of a unique gift.

    Attributes:
        name: The name of the model.
        sticker: The sticker that represents the unique gift.
        rarity_per_mille: The number of unique gifts that receive this model
            for every 1000 gifts upgraded.
        rarity: Rarity of the model if it is a crafted model, one of
            ``uncommon``, ``rare``, ``epic`` or ``legendary``.
    """

    name: str
    sticker: Sticker
    rarity_per_mille: int
    rarity: str | None = None


@dataclasses.dataclass(frozen=True, slots=True)
class UniqueGiftSymbol(TelegramObject):
    """The symbol displayed on a unique gift backdrop.

    Attributes:
        name: The name of the symbol.
        sticker: The sticker that represents the unique gift.
        rarity_per_mille: The number of unique gifts that receive this model
            for every 1000 gifts upgraded.
    """

    name: str
    sticker: Sticker
    rarity_per_mille: int


@dataclasses.dataclass(frozen=True, slots=True)
class UniqueGiftColors(TelegramObject):
    """Colors representing a unique gift in the interface.

    Attributes:
        model_custom_emoji_id: Custom emoji identifier of the unique gift's
            model.
        symbol_custom_emoji_id: Custom emoji identifier of the unique gift's
            symbol.
        light_theme_main_color: Main color used in light themes, in RGB format.
        light_theme_other_colors: List of 1-3 additional colors used in light
            themes, in RGB format.
        dark_theme_main_color: Main color used in dark themes, in RGB format.
        dark_theme_other_colors: List of 1-3 additional colors used in dark
            themes, in RGB format.
    """

    model_custom_emoji_id: str
    symbol_custom_emoji_id: str
    light_theme_main_color: int
    light_theme_other_colors: list[int]
    dark_theme_main_color: int
    dark_theme_other_colors: list[int]


@dataclasses.dataclass(frozen=True, slots=True)
class UniqueGift(TelegramObject):
    """This object represents a unique gift.

    Attributes:
        gift_id: Identifier of the regular gift from which this unique gift was
            upgraded.
        base_name: Human-readable name of the regular gift; may be empty if the
            gift was not upgraded.
        name: Unique name of the gift.
        number: Unique number of the upgraded gift.
        model: The model of the gift.
        symbol: The symbol of the gift.
        backdrop: The backdrop of the gift.
        is_premium: True, if the original regular gift was exclusively
            purchaseable by Telegram Premium subscribers.
        is_burned: True, if the gift was used to craft another gift and isn't
            available anymore.
        is_from_blockchain: True, if the gift is assigned from the TON
            blockchain.
        colors: The color scheme for user's name, replies, etc.
        publisher_chat: Information about the chat that published the gift.
    """

    gift_id: str
    base_name: str
    name: str
    number: int
    model: UniqueGiftModel
    symbol: UniqueGiftSymbol
    backdrop: UniqueGiftBackdrop
    is_premium: bool | None = None
    is_burned: bool | None = None
    is_from_blockchain: bool | None = None
    colors: UniqueGiftColors | None = None
    publisher_chat: Chat | None = None


@dataclasses.dataclass(frozen=True, slots=True)
class UniqueGiftInfo(TelegramObject):
    """Information about a unique gift attached to an owned gift.

    Attributes:
        gift: Information about the gift.
        origin: Origin of the gift, one of ``upgrade``, ``transfer``,
            ``resale``, ``gifted_upgrade`` or ``offer``.
        text: Text of the message that was added to the gift.
        entities: Special entities that appear in the text.
        is_private: True, if the sender and gift text are shown only to the
            gift receiver.
        last_resale_currency: Currency in which the payment for the gift was
            done, ``XTR`` or ``TON``.
        last_resale_amount: Price paid for the gift in either Telegram Stars or
            nanograms.
        owned_gift_id: Unique identifier of the received gift for the bot;
            only present if the bot has a relationship with the gift.
        transfer_star_count: Number of Telegram Stars that must be paid to
            transfer the gift; omitted if the bot cannot transfer the gift.
        next_transfer_date: Point in time (Unix timestamp) when the gift can be
            transferred.
    """

    gift: UniqueGift
    origin: str
    text: str | None = None
    entities: list[MessageEntity] | None = None
    is_private: bool | None = None
    last_resale_currency: str | None = None
    last_resale_amount: int | None = None
    owned_gift_id: str | None = None
    transfer_star_count: int | None = None
    next_transfer_date: int | None = None


@dataclasses.dataclass(frozen=True, slots=True)
class OwnedGiftRegular(TelegramObject):
    """A regular (non-unique) gift owned by a user or chat.

    Attributes:
        type: Type of the owned gift, always ``regular``.
        gift: Information about the gift.
        send_date: Point in time (Unix timestamp) when the gift was sent.
        owned_gift_id: Unique identifier of the received gift for the bot;
            only present if the bot has a relationship with the sender.
        sender_user: Identifier of the user that sent the gift, if any.
        text: Text of the message that was added to the gift.
        entities: Special entities that appear in the text.
        is_private: True, if the sender and gift text are shown only to the
            gift receiver.
        is_saved: True, if the gift is saved by the recipient.
        can_be_upgraded: True, if the gift can be upgraded to a unique gift.
        was_refunded: True, if the gift was refunded and isn't available
            anymore.
        convert_star_count: Number of Telegram Stars that can be claimed by the
            receiver instead of sending the gift.
        prepaid_upgrade_star_count: Number of Telegram Stars that were paid by
            the sender from the bot's balance to make the upgrade free for the
            receiver.
        is_upgrade_separate: True, if the gift can be upgraded to a unique gift
            only after it is converted to stars.
        unique_gift_number: Unique number of the upgraded gift the regular gift
            was converted from.
    """

    type: str
    gift: Gift
    send_date: int
    owned_gift_id: str | None = None
    sender_user: User | None = None
    text: str | None = None
    entities: list[MessageEntity] | None = None
    is_private: bool | None = None
    is_saved: bool | None = None
    can_be_upgraded: bool | None = None
    was_refunded: bool | None = None
    convert_star_count: int | None = None
    prepaid_upgrade_star_count: int | None = None
    is_upgrade_separate: bool | None = None
    unique_gift_number: int | None = None

    _DISCRIMINATOR = ("type", "regular")


@dataclasses.dataclass(frozen=True, slots=True)
class OwnedGiftUnique(TelegramObject):
    """A unique collectible gift owned by a user or chat.

    Attributes:
        type: Type of the owned gift, always ``unique``.
        gift: Information about the gift.
        send_date: Point in time (Unix timestamp) when the gift was sent.
        owned_gift_id: Unique identifier of the received gift for the bot;
            only present if the bot has a relationship with the sender.
        sender_user: Identifier of the user that sent the gift, if any.
        is_saved: True, if the gift is saved by the recipient.
        can_be_transferred: True, if the gift can be transferred to another
            owner.
        transfer_star_count: Number of Telegram Stars that must be paid to
            transfer the gift; omitted if the bot cannot transfer the gift.
        next_transfer_date: Point in time (Unix timestamp) when the gift can be
            transferred.
    """

    type: str
    gift: UniqueGift
    send_date: int
    owned_gift_id: str | None = None
    sender_user: User | None = None
    is_saved: bool | None = None
    can_be_transferred: bool | None = None
    transfer_star_count: int | None = None
    next_transfer_date: int | None = None

    _DISCRIMINATOR = ("type", "unique")


#: A single entry of an :class:`OwnedGifts` list — regular or unique.
OwnedGift = OwnedGiftRegular | OwnedGiftUnique


@dataclasses.dataclass(frozen=True, slots=True)
class OwnedGifts(TelegramObject):
    """A list of gifts owned by a user, a chat, or a business account.

    Returned by ``Bot.get_user_gifts``, ``Bot.get_chat_gifts`` and
    ``Bot.get_business_account_gifts``.

    Attributes:
        total_count: The total number of gifts owned by the user or chat that
            these gifts belong to; may be greater than the number of gifts in
            the list.
        gifts: The list of gifts.
        next_offset: Offset for the next request; if omitted, then this is the
            final result.
    """

    total_count: int
    gifts: list[OwnedGift]
    next_offset: str | None = None

File: packages/go/pkg/types/gifts.go
package types

// GiftBackground describes the background of a gift.
//
// See https://core.telegram.org/bots/api#giftbackground
type GiftBackground struct {
	// Center color of the background in RGB format.
	CenterColor int `json:"center_color"`
	// Edge color of the background in RGB format.
	EdgeColor int `json:"edge_color"`
	// Text color of the background in RGB format.
	TextColor int `json:"text_color"`
}

// Gift represents a gift that can be sent by the bot.
//
// See https://core.telegram.org/bots/api#gift
type Gift struct {
	// Unique identifier of the gift.
	ID string `json:"id"`
	// The sticker that represents the gift.
	Sticker Sticker `json:"sticker"`
	// The number of Telegram Stars that must be paid to send the sticker.
	StarCount int `json:"star_count"`
	// Number of Telegram Stars that must be paid to upgrade the gift to a unique one.
	UpgradeStarCount int `json:"upgrade_star_count,omitempty"`
	// True, if the gift can only be purchased by Telegram Premium subscribers.
	IsPremium bool `json:"is_premium,omitempty"`
	// True, if the gift can be used (after being upgraded) to customize a user's appearance.
	HasColors bool `json:"has_colors,omitempty"`
	// The total number of the gifts of this type that can be sent by all users; for limited gifts only.
	TotalCount int `json:"total_count,omitempty"`
	// The number of remaining gifts of this type that can be sent by all users; for limited gifts only.
	RemainingCount int `json:"remaining_count,omitempty"`
	// The total number of gifts of this type that can be sent by the bot; for limited gifts only.
	PersonalTotalCount int `json:"personal_total_count,omitempty"`
	// The number of remaining gifts of this type that can be sent by the bot; for limited gifts only.
	PersonalRemainingCount int `json:"personal_remaining_count,omitempty"`
	// Background of the gift.
	Background *GiftBackground `json:"background,omitempty"`
	// The total number of different unique gifts that can be obtained by upgrading the gift.
	UniqueGiftVariantCount int `json:"unique_gift_variant_count,omitempty"`
	// Information about the chat that published the gift.
	PublisherChat *Chat `json:"publisher_chat,omitempty"`
}

// Gifts represents a list of gifts.
//
// See https://core.telegram.org/bots/api#gifts
type Gifts struct {
	// The list of gifts.
	Gifts []Gift `json:"gifts"`
}

// UniqueGiftBackdropColors describes the colors of a unique gift backdrop.
//
// See https://core.telegram.org/bots/api#uniquegiftbackdropcolors
type UniqueGiftBackdropColors struct {
	// The color in the center of the backdrop in RGB format.
	CenterColor int `json:"center_color"`
	// The color on the edges of the backdrop in RGB format.
	EdgeColor int `json:"edge_color"`
	// The color to be applied to the symbol in RGB format.
	SymbolColor int `json:"symbol_color"`
	// The color for the text on the backdrop in RGB format.
	TextColor int `json:"text_color"`
}

// UniqueGiftBackdrop describes the backdrop of a unique gift.
//
// See https://core.telegram.org/bots/api#uniquegiftbackdrop
type UniqueGiftBackdrop struct {
	// Name of the backdrop.
	Name string `json:"name"`
	// Colors of the backdrop.
	Colors UniqueGiftBackdropColors `json:"colors"`
	// The number of unique gifts that receive this backdrop for every 1000 gifts upgraded.
	RarityPerMille int `json:"rarity_per_mille"`
}

// UniqueGiftModel describes the model of a unique gift.
//
// See https://core.telegram.org/bots/api#uniquegiftmodel
type UniqueGiftModel struct {
	// Name of the model.
	Name string `json:"name"`
	// The sticker that represents the unique gift.
	Sticker Sticker `json:"sticker"`
	// The number of unique gifts that receive this model for every 1000 gift upgrades.
	RarityPerMille int `json:"rarity_per_mille"`
	// Rarity of the model if it is a crafted model; one of "uncommon", "rare",
	// "epic" or "legendary".
	Rarity string `json:"rarity,omitempty"`
}

// UniqueGiftSymbol describes the symbol of a unique gift.
//
// See https://core.telegram.org/bots/api#uniquegiftsymbol
type UniqueGiftSymbol struct {
	// Name of the symbol.
	Name string `json:"name"`
	// The sticker that represents the unique gift.
	Sticker Sticker `json:"sticker"`
	// The number of unique gifts that receive this model for every 1000 gifts upgraded.
	RarityPerMille int `json:"rarity_per_mille"`
}

// UniqueGiftColors describes the color scheme of a unique gift.
//
// See https://core.telegram.org/bots/api#uniquegiftcolors
type UniqueGiftColors struct {
	// Custom emoji identifier of the unique gift's model.
	ModelCustomEmojiID string `json:"model_custom_emoji_id"`
	// Custom emoji identifier of the unique gift's symbol.
	SymbolCustomEmojiID string `json:"symbol_custom_emoji_id"`
	// Main color used in light themes; RGB format.
	LightThemeMainColor int `json:"light_theme_main_color"`
	// List of 1-3 additional colors used in light themes; RGB format.
	LightThemeOtherColors []int `json:"light_theme_other_colors"`
	// Main color used in dark themes; RGB format.
	DarkThemeMainColor int `json:"dark_theme_main_color"`
	// List of 1-3 additional colors used in dark themes; RGB format.
	DarkThemeOtherColors []int `json:"dark_theme_other_colors"`
}

// UniqueGift describes a unique collectible gift with a potentially upgradable model.
//
// See https://core.telegram.org/bots/api#uniquegift
type UniqueGift struct {
	// Identifier of the regular gift from which this gift was upgraded.
	GiftID string `json:"gift_id"`
	// Human-readable name of the regular gift.
	BaseName string `json:"base_name"`
	// Unique name of the gift.
	Name string `json:"name"`
	// Unique number of the upgraded gift.
	Number int `json:"number"`
	// Model of the gift.
	Model UniqueGiftModel `json:"model"`
	// Symbol of the gift.
	Symbol UniqueGiftSymbol `json:"symbol"`
	// Backdrop of the gift.
	Backdrop UniqueGiftBackdrop `json:"backdrop"`
	// True, if the original regular gift was exclusively purchaseable by Telegram Premium subscribers.
	IsPremium bool `json:"is_premium,omitempty"`
	// True, if the gift was used to craft another gift and isn't available anymore.
	IsBurned bool `json:"is_burned,omitempty"`
	// True, if the gift is assigned from the TON blockchain.
	IsFromBlockchain bool `json:"is_from_blockchain,omitempty"`
	// The color scheme for the user's name, replies, etc.
	Colors *UniqueGiftColors `json:"colors,omitempty"`
	// Information about the chat that published the gift.
	PublisherChat *Chat `json:"publisher_chat,omitempty"`
}

// UniqueGiftInfo describes information about a unique gift.
//
// See https://core.telegram.org/bots/api#uniquegiftinfo
type UniqueGiftInfo struct {
	// Information about the gift.
	Gift UniqueGift `json:"gift"`
	// Origin of the gift; one of "upgrade", "transfer", "resale",
	// "gifted_upgrade" or "offer".
	Origin string `json:"origin"`
	// Text of the message that was added to the gift.
	Text string `json:"text,omitempty"`
	// Special entities that appear in the text.
	Entities []MessageEntity `json:"entities,omitempty"`
	// True, if the sender and gift text are shown only to the gift receiver.
	IsPrivate bool `json:"is_private,omitempty"`
	// Currency in which the payment for the gift was done; "XTR" or "TON".
	LastResaleCurrency string `json:"last_resale_currency,omitempty"`
	// Price paid for the gift in either Telegram Stars or nanograms.
	LastResaleAmount int `json:"last_resale_amount,omitempty"`
	// Unique identifier of the received gift for the bot.
	OwnedGiftID string `json:"owned_gift_id,omitempty"`
	// Number of Telegram Stars that must be paid to transfer the gift.
	TransferStarCount int `json:"transfer_star_count,omitempty"`
	// Point in time (Unix timestamp) when the gift can be transferred.
	NextTransferDate int64 `json:"next_transfer_date,omitempty"`
}

// GiftPremiumSubscriptionOptions represents parameters for the giftPremiumSubscription method.
//
// Telegram API: https://core.telegram.org/bots/api#giftpremiumsubscription
type GiftPremiumSubscriptionOptions struct {
	// Unique identifier of the target user.
	UserID int64 `json:"user_id"`
	// Number of months the subscription will be gifted; must be one of 3, 6, 12.
	MonthCount int `json:"month_count"`
	// Number of Telegram Stars to be charged; must be 1000, 1500 or 2500 for 3, 6, 12 months respectively.
	StarCount int `json:"star_count"`
	// Text shown to the recipient alongside the payment receipt; 0-255 characters after entities parsing.
	Text string `json:"text,omitempty"`
	// Mode for parsing entities in the text.
	TextParseMode string `json:"text_parse_mode,omitempty"`
	// A JSON-serialized list of special entities that appear in the text.
	TextEntities []MessageEntity `json:"text_entities,omitempty"`
}
